File: app/src/main/java/com/hex/evegate/ui/MainActivity.java
package com.hex.evegate.ui;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;
import com.hex.evegate.R;
import com.hex.evegate.api.StationApi;
import com.hex.evegate.api.dto.NowPlayingDto;
import com.hex.evegate.net.RetrofitClient;
import com.hex.evegate.radio.PlaybackStatus;
import com.hex.evegate.radio.RadioManager;
import com.hex.evegate.util.Shoutcast;
import com.hex.evegate.util.ShoutcastHelper;
import com.hex.evegate.util.ShoutcastListAdapter;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;
import retrofit2.Retrofit;

public class MainActivity extends AppCompatActivity {

    ImageButton trigger;
    ListView listView;
    TextView tvCount;
    TextView tvSongName;
    ImageView ivImage;
    ImageView ivBackground;
    TextView textView;
    LinearLayout llFirst;
    TextView tvFirst;
    ImageView ivFirst;
    LinearLayout llSecond;
    TextView tvSecond;
    ImageView ivSecond;
    View subPlayer;

    RadioManager radioManager;

    String streamURL;
    private long lastBackPressTime;

    private CompositeDisposable compositeDisposable;
    private Retrofit retrofit;
    private StationApi stationApi;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initialize();
        configNet();
        getNowPlayingDto();
    }

    private void initialize() {

        setContentView(R.layout.activity_main);

        radioManager = RadioManager.with(this);

        trigger = findViewById(R.id.playTrigger);
        listView = findViewById(R.id.listview);
        listView.setAdapter(new ShoutcastListAdapter(this, ShoutcastHelper.retrieveShoutcasts(this)));
        tvCount = findViewById(R.id.tvCount);
        tvSongName = findViewById(R.id.tvSongName);
        ivImage = findViewById(R.id.ivImage);
        ivBackground = findViewById(R.id.ivBackground);
        textView = findViewById(R.id.name);
        llFirst = findViewById(R.id.llFirst);
        tvFirst = findViewById(R.id.tvFirst);
        ivFirst = findViewById(R.id.ivFirst);
        llSecond = findViewById(R.id.llSecond);
        tvSecond = findViewById(R.id.tvSecond);
        ivSecond = findViewById(R.id.ivSecond);
        subPlayer = findViewById(R.id.sub_player);

        trigger.setOnClickListener(v -> {
            if(TextUtils.isEmpty(streamURL)) return;

            radioManager.playOrPause(streamURL);
        });

        listView.setOnItemClickListener((parent, view, position, id) -> {
            Shoutcast shoutcast = (Shoutcast) parent.getItemAtPosition(position);
            if(shoutcast == null){

                return;

            }

            textView.setText(shoutcast.getName());

            subPlayer.setVisibility(View.VISIBLE);

            streamURL = shoutcast.getUrl();

            radioManager.playOrPause(streamURL);
        });

        llFirst.setOnClickListener(firstClickListener);
        tvFirst.setOnClickListener(firstClickListener);
        ivFirst.setOnClickListener(firstClickListener);

        llSecond.setOnClickListener(secondClickListener);
        tvSecond.setOnClickListener(secondClickListener);
        ivSecond.setOnClickListener(secondClickListener);
    }

    private void getNowPlayingDto() {
        compositeDisposable.add(stationApi.nowPlaying()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(this::handleNowPlayingResponse, this::handleNowPlayingError));
    }

    private void handleNowPlayingResponse(Response<NowPlayingDto> result) {
        if (result.isSuccessful()) {
            if (result.body() != null) {
                tvCount.setText(result.body().getListeners().getTotal());
                tvSongName.setText(result.body().getNow_playing().getSong().getText());
                try {
                    Glide.with(this).load(result.body().getNow_playing().getSong().getArt())
                            .into(ivBackground);
                } catch (Exception e) {
                    /*ignored*/
                }
            }
        } else {
            Toast.makeText(this, "Ашипко!", Toast.LENGTH_SHORT).show();
        }
    }

    private void handleNowPlayingError(Throwable error) {
        Toast.makeText(this, "Ашипко!", Toast.LENGTH_SHORT).show();
    }

    private void configNet() {
        compositeDisposable = new CompositeDisposable();
        retrofit = RetrofitClient.INSTANCE.getInstance();
        stationApi = retrofit.create(StationApi.class);
    }

    View.OnClickListener firstClickListener = v -> startActivity(new Intent(Intent.ACTION_VIEW,
            Uri.parse("https://t.me/evegateradio")));

    View.OnClickListener secondClickListener = v -> startActivity(new Intent(Intent.ACTION_VIEW,
            Uri.parse("https://t.me/EVE_ONLINE_RUS")));

    @Override
    public void onStart() {

        super.onStart();

        EventBus.getDefault().register(this);
    }

    @Override
    public void onStop() {

        EventBus.getDefault().unregister(this);

        super.onStop();
    }

    @Override
    protected void onDestroy() {

        radioManager.unbind();
        compositeDisposable.clear();

        super.onDestroy();
    }

    @Override
    protected void onResume() {
        super.onResume();

        radioManager.bind();
    }

    @Override
    public void onBackPressed() {
        long now = System.currentTimeMillis();
        if (System.currentTimeMillis() - lastBackPressTime < 1000) {
            finish();
        } else {
            lastBackPressTime = now;
            Toast toast = Toast.makeText(MainActivity.this, "Нажмите \"Назад\" еще раз для выхода их приложения.\nДля продолжения прослушивания в фоновом режиме нажмите \"Свернуть\"", Toast.LENGTH_LONG);
            toast.setGravity(Gravity.CENTER, 0, 0);
            toast.show();
        }
    }

    @Subscribe
    public void onEvent(String status){

        switch (status){

            case PlaybackStatus.LOADING:

                // loading

                break;

            case PlaybackStatus.ERROR:

                Toast.makeText(this, R.string.no_stream, Toast.LENGTH_SHORT).show();

                break;

        }

        trigger.setImageResource(status.equals(PlaybackStatus.PLAYING)
                ? R.drawable.ic_pause_black
                : R.drawable.ic_play_arrow_black);

    }
}

File: app/src/main/java/com/hex/evegate/ui/activity/MainActivity.kt
package com.hex.evegate.ui.activity

import android.content.Intent
import android.media.AudioManager
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import com.arellomobile.mvp.MvpAppCompatActivity
import com.arellomobile.mvp.presenter.InjectPresenter
import com.bumptech.glide.Glide
import com.google.android.material.navigation.NavigationView
import com.hex.evegate.R
import com.hex.evegate.radio.PlaybackStatus
import com.hex.evegate.radio.RadioManager
import com.hex.evegate.ui.mvp.presenter.MainPresenter
import com.hex.evegate.ui.mvp.view.MainView
import com.hex.evegate.ui.visualizer.BarVisualizer
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

class MainActivity : MvpAppCompatActivity(), MainView, NavigationView.OnNavigationItemSelectedListener {
    @InjectPresenter
    lateinit var mainPresenter: MainPresenter

    private lateinit var tvCount: TextView
    private lateinit var tvSongName: TextView
    private lateinit var tvPlaylist: TextView
    private lateinit var ivBackground: ImageView
    private lateinit var chbHQ: CheckBox
    private lateinit var ibPlayPause: ImageButton
    private lateinit var dlDrawer: DrawerLayout
    private lateinit var nvMenu: NavigationView
    private lateinit var llProgressStart: LinearLayout
    private lateinit var llProgressEnd: LinearLayout

    private var lastBackPressTime: Long = 0

    private var bvVisualizer: BarVisualizer? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mainPresenter.configNet()
        initialize()
    }

    override fun initialize() {
        setContentView(R.layout.activity_main)

        tvCount = findViewById(R.id.tvCount)
        tvSongName = findViewById(R.id.tvSongName)
        tvPlaylist = findViewById(R.id.tvPlaylist)
        ivBackground = findViewById(R.id.ivBackground)
        dlDrawer = findViewById(R.id.dlDrawer)
        nvMenu = findViewById(R.id.nvMenu)
        val toggle = ActionBarDrawerToggle(
                this, dlDrawer, findViewById(R.id.toolbar), R.string.on, R.string.off)
        dlDrawer.addDrawerListener(toggle)
        toggle.syncState()
        nvMenu.setNavigationItemSelectedListener(this)

        chbHQ = findViewById(R.id.chbHQ)
        chbHQ.isChecked = mainPresenter.isHQ()
        chbHQ.setOnCheckedChangeListener { _, isChecked ->
            mainPresenter.onCheckBoxHqChanged(isChecked)
        }

        ibPlayPause = findViewById(R.id.ibPlayPause)
        ibPlayPause.setOnClickListener {
            mainPresenter.playOrPause()
        }
        ibPlayPause.setImageResource(if (mainPresenter.isPlaying()) {
            android.R.drawable.ic_media_pause } else { android.R.drawable.ic_media_play }
        )

        llProgressStart = findViewById(R.id.llProgressStart)
        llProgressEnd = findViewById(R.id.llProgressEnd)
    }

    public override fun onStart() {
        super.onStart()

        EventBus.getDefault().register(this)
    }

    override fun startVisualizer() {
        if (mainPresenter.isPlaying()) {
            bvVisualizer?.visualizer?.release()
            volumeControlStream = AudioManager.STREAM_MUSIC
            val id = RadioManager.service?.exoPlayer?.audioSessionId
            if (id != null) {
                bvVisualizer = findViewById<BarVisualizer>(R.id.bvVisualizer).apply {
                    setColor(ContextCompat.getColor(this@MainActivity, R.color.barVisualizer))
                    setDensity(70F)
                    setPlayer(id)
                    visibility = View.VISIBLE
                }
            }
        }
    }

    override fun stopVisualizer() {
        bvVisualizer?.let {
            it.visibility = View.INVISIBLE
            it.release()
            it.visualizer?.enabled = false
        }
    }

    public override fun onStop() {

        EventBus.getDefault().unregister(this)

        super.onStop()
    }

    override fun onResume() {
        super.onResume()

        initialize()
        startVisualizer()
    }

    override fun onPause() {
        super.onPause()
        try {stopVisualizer()} catch (e: Exception) {}
    }

    override fun onBackPressed() {
        if (dlDrawer.isDrawerOpen(GravityCompat.START)) {
            dlDrawer.closeDrawer(GravityCompat.START)
        } else {
            val now = System.currentTimeMillis()
            if (System.currentTimeMillis() - lastBackPressTime < 1000) {
                finish()
            } else {
                lastBackPressTime = now
                showMessage("Нажмите \"Назад\" еще раз для выхода.")
            }
        }

    }

    @Subscribe
    override fun onEvent(status: String) {
        if (status == PlaybackStatus.PLAYING) {
            startVisualizer()
            ibPlayPause.setImageResource(android.R.drawable.ic_media_pause)
        } else {
            try {stopVisualizer()} catch (e: Exception) {}
            ibPlayPause.setImageResource(android.R.drawable.ic_media_play)
        }
        when (status) {
            PlaybackStatus.LOADING -> {
                ibPlayPause.setImageResource(R.drawable.ic_cloud_download_white_24dp)
            }
            PlaybackStatus.ERROR ->
                showMessage(R.string.no_stream)
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.history -> startActivity(Intent(this, HistoryActivity::class.java))
            R.id.community -> startActivity(Intent(this, CommsActivity::class.java))
        }
        dlDrawer.closeDrawer(GravityCompat.START)
        return true
    }

    override fun showProgress(percent: Float) {
        llProgressStart.layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT, percent)
        llProgressEnd.layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT, Math.abs((100 - percent)))
    }

    override fun showMessage(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    override fun showMessage(res: Int) {
        Toast.makeText(this, res, Toast.LENGTH_SHORT).show()
    }

    override fun showArt(artUrl: String) {
        try {
            Glide.with(this).load(artUrl).into(ivBackground)
        } catch (e: Exception) { /*ignored*/ }
    }

    override fun showLive(isLive: Boolean) {
        findViewById<ImageView>(R.id.ivLive).visibility =
                if (isLive) { ImageView.VISIBLE } else { ImageView.GONE }
    }

    override fun setPlayList(playlist: String) { tvPlaylist.text = playlist }
    override fun setSongName(songName: String) { tvSongName.text = songName }
    override fun setCount(count: String) { tvCount.text = count }
}
